### PR TITLE
Configure Ruff to apply flake8-bugbear/isort/pyupgrade

### LINKED ISSUE #1841: Ruff conflicts with pep8
In this PR #1839 I have a conflict where the @pep8speaks bot is in conflict with the ruff formatter.
PEP8 comment: https://github.com/zarr-developers/zarr-python/pull/1839#issuecomment-2094937391
Ruff pre-commit: https://results.pre-commit.ci/run/github/48049137/1715025400.r2onHl8LSX240CJ0_K4lOw

The difference is illustrated in this commit: https://github.com/zarr-developers/zarr-python/pull/1839/commits/d8b5ca7748f0d1a0bcb049aa414aac4bdbc2d876

### PR DESCRIPTION
**Pull request by jhamman**
Fixes #1841 
Closes #1862 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by DimitriPapadopoulos** [comment]
See also existing pull requests #1703, #1704.

**Comment by DimitriPapadopoulos** [comment]
Why apply `ruff formatting`? As far as I can tell, this project is currently using `black` as a formatter:

https://github.com/zarr-developers/zarr-python/blob/cb4230dd9a2a0c68b78384b24a2a5567df0b704a/.pre-commit-config.yaml#L14-L17

**Comment by jhamman** [comment]
@DimitriPapadopoulos - this PR is pointing at the `v3` branch where we aren't using black at the moment. I don't see any problem continuing with https://github.com/zarr-developers/zarr-python/pull/1702, https://github.com/zarr-developers/zarr-python/pull/1703, and https://github.com/zarr-developers/zarr-python/pull/1704 but those are not going to make it into this branch for the 3.0 release. See https://github.com/zarr-developers/zarr-python/issues/1777 for some additional details on our plan for the `v3` branch.

**Comment by DimitriPapadopoulos** [comment]
Understood. So if some additional ruff rules are eventually accepted in the [`main`](https://github.com/zarr-developers/zarr-python/tree/main) branch (such as `ISC`, `RSE` or `RUF`), I'll have to manually create a distinct pull request against branch [`v3`](https://github.com/zarr-developers/zarr-python/tree/v3) so that they end up in 3.* series, won't I?

**Comment by jhamman** [comment]
> I'll have to manually create a distinct pull request against branch [v3](https://github.com/zarr-developers/zarr-python/tree/v3) so that they end up in 3.* series, won't I?

That's right.

BTW, I know you've been working on this topic here in Zarr land for a while. Now that I see your PRs against the V3 branch, I'm happy to step aside and let you run that to completion. Would you like me to close this PR?

**Comment by DimitriPapadopoulos** [comment]
Not at all, please proceed with this PR. I have added a few additional rules that seem of interest to me: #1867, #1868, #1869 against [v3](https://github.com/zarr-developers/zarr-python/tree/v3). I will rebase them on top of your PR once you're done.

File: pyproject.toml [hunk 1]
```diff
@@ -160,7 +160,15 @@ extend-exclude = [
     "build",
     "dist",
     "venv",
-    "docs"
+    "docs",
+    "src/zarr/v2/",
+    "tests/v2/"
```
Comment (jhamman) [inline pyproject.toml:165]: this exclusion didn't seem to work. I'll rerun.

**Comment by jhamman** [comment]
This is likely to cause some serious merge conflicts with #1857 - I'm going to hold on this until that has been merged. Hopefully tomorrow 🐎 .

**Comment by normanrz** [comment]
How much manual work did you have to do for this? If most is automated, I am wondering if just redo the automated steps once #1670 and #1857 are merged? Otherwise, I'm happy to resolve the conflicts.

File: src/zarr/codecs/sharding.py [hunk 2]
```diff
@@ -615,7 +619,7 @@ def _shard_index_size(self, chunks_per_shard: ChunkCoords) -> int:
             16 * product(chunks_per_shard), self._get_index_chunk_spec(chunks_per_shard)
         )
 
-    @lru_cache
+    @lru_cache  # noqa: B019
```
Comment (normanrz) [inline src/zarr/codecs/sharding.py:622]: Good to know! Fixed in https://github.com/zarr-developers/zarr-python/pull/1670/commits/530e88b6cd63d4d91c44302f0bc0c6a37749bee4

**Review by normanrz — APPROVED** [review]

**Comment by jhamman** [comment]
> How much manual work did you have to do for this? If most is automated, I am wondering if just redo the automated steps once https://github.com/zarr-developers/zarr-python/pull/1670 and https://github.com/zarr-developers/zarr-python/pull/1857 are merged? Otherwise, I'm happy to resolve the conflicts.

Little manual work went in here. I'm happy to reapply after your PRs go in.

**Comment by normanrz** [comment]
Closing in favor of https://github.com/zarr-developers/zarr-python/pull/1890

**Pull request closed without merging**